spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/ExpansionStrategy.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.completion;
+
+import java.util.List;
+
+import org.springframework.cloud.dataflow.core.StreamDefinition;
+
+/**
+ * Used to enhance a well formed stream definition by adding yet more text to it (<i>e.g.</i> adding more options to a
+ * module).
+ * 
+ * @author Eric Bottard
+ */
+public interface ExpansionStrategy {
```

Comment:
You're right about the two invocation thing (which is not uncommon in other designs). This can be avoided without changing the method signature though.

And indeed, I think I'll keep the "mutate the list" approach, because I will need the return value (currently `null`) to indicate whether or not a Strategy should short-circuit the whole invocation chain. I discovered it will be needed for cases like ` ... | log --level=INF<TAB>` which complete with a well known value (from an enum) and so we *know* for sure at the semantic level that `INF` is not sufficient. Yet, at the pure lexical parsing level, `--level=INF` was wellformed, and so other strategies get invoked, resulting in propositions like `--level=INF --name=` (in addition to `--level=INFO` which is the only one we want, at that time). 

Hope that makes sense.